bigchaindb/upsert_validator/validator_election.py

```diff
@@ -0,0 +1,140 @@
+from bigchaindb.common.exceptions import (InvalidSignature,
+                                          DuplicateTransaction)
+from bigchaindb.tendermint.utils import key_from_base64
+from bigchaindb.common.crypto import (public_key_from_ed25519_key)
+from bigchaindb.common.transaction import Transaction, Input, Output
+from bigchaindb.common.schema import (_load_schema,
+                                      _validate_schema,
+                                      TX_SCHEMA_VERSION,
+                                      TX_SCHEMA_COMMON,
+                                      TX_SCHEMA_CREATE)
+
+
+_, TX_SCHEMA_VALIDATOR_ELECTION = _load_schema('transaction_validator_election_' +
+                                               TX_SCHEMA_VERSION, __file__)
+
+
+class ValidatorElection(Transaction):
+
+    VALIDATOR_ELECTION = 'VALIDATOR_ELECTION'
+    # NOTE: this transaction class extends create so the operation inheritence is achieved
+    # by renaming CREATE to VALIDATOR_ELECTION
+    CREATE = VALIDATOR_ELECTION
+    ALLOWED_OPERATIONS = (VALIDATOR_ELECTION,)
+
+    def __init__(self, operation, asset, inputs=None, outputs=None,
+                 metadata=None, version=None, hash_id=None):
+        # operation `CREATE` is being passed as argument as `VALIDATOR_ELECTION` is an extension
+        # of `CREATE` and any validation on `CREATE` in the parent class should apply to it
+        super().__init__(operation, asset, inputs, outputs, metadata, version, hash_id)
+
+        self.operation = self.VALIDATOR_ELECTION
+
+    @classmethod
+    def current_validators(cls, bigchain):
+        """Return a dictionary of validators with key as `public_key` and
+           value as the `voting_power`
+        """
+
+        validators = {}
+        for validator in bigchain.get_validators():
+            # NOTE: we assume that Tendermint encodes public key in base64
+            public_key = public_key_from_ed25519_key(key_from_base64(validator['pub_key']['value']))
+            validators[public_key] = validator['voting_power']
+
+        return validators
+
+    @classmethod
+    def recipients(cls, bigchain):
+        """Convert validator dictionary to a recipient list for `Transaction`"""
+
+        recipients = []
+        for public_key, voting_power in cls.current_validators(bigchain).items():
+            recipients.append(([public_key], voting_power))
+
+        return recipients
+
+    @classmethod
+    def is_same_topology(cls, current_topology, election_topology):
+        voters = {}
+        for voter in election_topology:
+            if len(voter.public_keys) > 1:
+                return False
+
+            [public_key] = voter.public_keys
+            voting_power = voter.amount
+            voters[public_key] = voting_power
+
+        # Check whether the voters and their votes is same to that of the
+        # validators and their voting power in the network
+        return (current_topology.items() == voters.items())
```

Comment:
Although the new Python guarantees the key order, I would compare dicts directly `current_topology == voters`, for simplicity.